subsys/net/ip/tcp2.c

```diff
@@ -0,0 +1,1584 @@
+		}
+		if (fl & ACK) {
+			s += snprintf(s, buf_size, "ACK=%u,", th_ack(th));
+			buf_size -= s - buf;
+		}
+		if (fl & PSH) {
+			s += snprintf(s, buf_size, "PSH,");
+			buf_size -= s - buf;
+		}
+		if (fl & RST) {
+			s += snprintf(s, buf_size, "RST,");
+			buf_size -= s - buf;
+		}
+		if (fl & URG) {
+			s += snprintf(s, buf_size, "URG,");
+			buf_size -= s - buf;
+		}
+		s[strlen(s) - 1] = '\0';
+		s--;
+	}
+
+	if (data_len) {
+		s += snprintf(s, buf_size, ", len=%zd", data_len);
+		buf_size -= s - buf;
+	}
+
+	if (((bool)(PSH & fl)) != (data_len > 0)) {
+		NET_WARN("Invalid TCP packet: %s, data_len=%zd", buf, data_len);
+	}
+end:
+	return buf;
+#undef BUF_SIZE
+}
+
+static void tcp_send(struct net_pkt *pkt)
+{
+	NET_DBG("%s", tcp_th(pkt));
+
+	tcp_pkt_ref(pkt);
+
+	if (tcp_send_cb) {
+		if (tcp_send_cb(pkt) < 0) {
+			NET_ERR("net_send_data()");
+			tcp_pkt_unref(pkt);
+		}
+		goto out;
+	}
+
+	if (net_send_data(pkt) < 0) {
+		NET_ERR("net_send_data()");
+		tcp_pkt_unref(pkt);
+	}
+out:
+	tcp_pkt_unref(pkt);
+}
+
+static void tcp_send_queue_flush(struct tcp *conn)
+{
+	struct net_pkt *pkt;
+
+	if (is_timer_subscribed(&conn->send_timer)) {
+		k_timer_stop(&conn->send_timer);
+	}
+
+	while ((pkt = tcp_slist(&conn->send_queue, get,
+				struct net_pkt, next))) {
+		tcp_pkt_unref(pkt);
+	}
+}
+
+static void tcp_win_free(struct tcp_win *w)
+{
+	struct net_buf *buf;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&w->bufs, buf, next) {
+		NET_DBG("%s %p len=%d", w->name, buf, buf->len);
+		tcp_nbuf_unref(buf);
+	}
+
+	tcp_free(w->name);
+	tcp_free(w);
+}
+
+int net_tcp_unref(struct net_context *context)
+{
+	struct tcp *conn = context->tcp;
+
+	NET_DBG("%p", context);
+
+	tp_out(conn->iface, "TP_TRACE", "event", "CONN_DELETE");
+
+	if (_tcp_conn_delete == false) {
+		goto out;
+	}
+
+	tcp_send_queue_flush(conn);
+
+	tcp_win_free(conn->snd);
+	tcp_win_free(conn->rcv);
+
+	tcp_free(conn->src);
+	tcp_free(conn->dst);
+
+	{
+		int key = irq_lock();
+
+		context->tcp = NULL;
+		memset(conn, 0, sizeof(*conn));
+		sys_slist_find_and_remove(&tcp_conns, (sys_snode_t *) conn);
+		k_mem_slab_free(&tcp_conns_slab, (void **)&conn);
+
+		irq_unlock(key);
+	}
+out:
+	return 0;
+}
+
+static void tcp_send_process(struct k_timer *timer)
+{
+	struct tcp *conn = k_timer_user_data_get(timer);
+	struct net_pkt *pkt = tcp_slist(&conn->send_queue, peek_head,
+					struct net_pkt, next);
+
+	NET_DBG("%s %s", tcp_th(pkt), conn->in_retransmission ?
+		"in_retransmission" : "");
+
+	if (conn->in_retransmission) {
+		if (conn->send_retries > 0) {
+			tcp_send(tcp_pkt_clone(pkt));
+			conn->send_retries--;
+		} else {
+			net_tcp_unref(conn->context);
+			conn = NULL;
+		}
+	} else {
+		u8_t fl = th_get(pkt)->th_flags;
+		bool forget = ACK == fl || PSH == fl || (ACK | PSH) == fl ||
+			RST & fl;
+
+		pkt = forget ? tcp_slist(&conn->send_queue, get, struct net_pkt,
+						next) : tcp_pkt_clone(pkt);
+		tcp_send(pkt);
+
+		if (false == forget && false == is_timer_subscribed(
```

Comment:
We usually write the comparisons so that the variable is first and the const is last.